core/extensions.js

```diff
@@ -75,6 +75,85 @@ Blockly.Extensions.apply = function(name, block) {
 };
 
 /**
+ * Builds an extension function that will map a dropdown value to a tooltip string.
+ * Tooltip strings will be passed through Blockly.utils.checkMessageReferences(..)
+ * immediately and Blockly.utils.replaceMessageReferences(..) at display time.
+ * @param {string} dropdownName The name of the field whose value is the key
+ *     to the lookup table.
+ * @param {!Object<string, string>} lookupTable The table of field values to
+ *     tooltip text.
+ * @return {Function} The extension function.
+ */
+Blockly.Extensions.buildTooltipForDropdown = function(dropdownName, lookupTable) {
+  // List of block types already validated, to minimize duplicate warnings.
+  var blockTypesChecked = [];
+
+  // Validate message strings early.
+  for (var key in lookupTable) {
+    Blockly.utils.checkMessageReferences(lookupTable[key]);
+  }
+
+  /**
+   * The actual extension.
+   * @this {Blockly.Block}
+   */
+  return function() {
+    var thisBlock = this;
+
+    if (this.type && !blockTypesChecked.includes(this.type)) {
+      Blockly.Extensions.checkDropdownOptionsInTable_(
+        this, dropdownName, lookupTable);
+      blockTypesChecked.push(this.type);
+    }
+
+    this.setTooltip(function() {
+      var value = thisBlock.getFieldValue(dropdownName);
+      var tooltip = lookupTable[value];
+      if (tooltip == null) {
+        if (!blockTypesChecked.includes(thisBlock.type)) {
+          // Warn for missing values on generated tooltips
```

Comment:
A helper function to print a warning, largely concatenating strings seems excessive.  Collapsed one line.